Swift 5: adopt new Data.withUnsafeBytes API

alanzeino opened the pull request:
Unfortunately I wasn't able to use the `#if compiler(>=5.0)` flag since the project supports older swift versions where this directive is not available.

thomasvl commented:
> Unfortunately I wasn't able to use the #if compiler(>=5.0) flag since the project supports older swift versions where this directive is not available.

Why?  We already use directive like that with problems.

alanzeino commented:
> > Unfortunately I wasn't able to use the #if compiler(>=5.0) flag since the project supports older swift versions where this directive is not available.
> 
> Why? We already use directive like that without problems.

Sorry I should've been more clear. `#if swift()` has been supported for a while but `#if compiler()` less so. The project at first glance supports Swift versions right back to 3.0, but `#if compiler()` was added much later than that.

I didn't see any in the codebase when I grep'd for it, but I would prefer to be wrong because I want to use that here :)

tbkka commented:
> The project at first glance supports Swift versions right back to 3.0,...

Our policy is to support one full major version back.  Since 5.0 is the current release version, that means we still support 4.0 and later.  Once 5.2 is released, we'll be able to use `#if compiler` since we'll only support >= 4.2 at that point.

thomasvl commented:
I also don't know the `#compiler` would be correct, as some of these seem to be API changes. API changes should be tied to the language version, not the underlying compiler version/release.  If they were tied to the compiler doesn't that mean some code that compiles with Swift 4.x would suddenly stop compiling in 4.x mode with the 5 compiler?

ps - I completely read that initial comment as `if #swift()`, adding to the confusion here. 😄

alanzeino commented:
> I also don't know the `#compiler` would be correct, as some of these seem to be API changes. API changes should be tied to the language version, not the underlying compiler version/release. If they were tied to the compiler doesn't that mean some code that compiles with Swift 4.x would suddenly stop compiling in 4.x mode with the 5 compiler?
> 
> ps - I completely read that initial comment as `#if swift()`, adding to the confusion here. 😄

Yep you're totally right. These changes are all Swift 5 compiler in 5.0 mode, not backwards compatible with 5 compiler in 4.2 mode.

tbkka commented on Sources/SwiftProtobuf/AnyMessageStorage.swift:
```diff
@@ -51,7 +51,11 @@ fileprivate func emitVerboseTextForm(visitor: inout TextFormatEncodingVisitor, m
 fileprivate func asJSONObject(body: Data) -> Data {
   let asciiOpenCurlyBracket = UInt8(ascii: "{")
   let asciiCloseCurlyBracket = UInt8(ascii: "}")
+  #if swift(>=5.0)
```
Is the conditional really needed here?

I just tried the new version `Data([asciiOpenCurlyBracket])` in Swift 4.0 and it seems to work correctly there.

tbkka commented on Sources/SwiftProtobuf/Data+Extensions.swift:
```diff
@@ -0,0 +1,49 @@
+// Sources/SwiftProtobuf/Data+Extensions.swift - Extension wrapping deprecated Data methods
```
Elsewhere, we've done this the other way around.

That is, update the code to use the _new_ syntax, and write an adapter to simulate the new syntax on old Swift versions.  It's a bit more work, but means that when we stop supporting those old versions, we can simply delete the adapter code.

tbkka replied:
If you could get the adapter right (simulating new syntax on old Swift), and prove that it works by converting just a few places, I'm happy to help get the rest converted.  I've gotten pretty handy at scripting global changes of this sort.

alanzeino replied:
So I took a look at it, and I think doing it in this fashion is maybe a bit worse than the current one. 

My opinion is based on this [tweet](https://twitter.com/jckarter/status/1071183859083960320) — basically Apple is saying that we should carefully check that the pointer is both valid and has a length (`count`).

For example, if we try to support the new syntax in the code, this call site (from `BinaryDelimited.swift`): 

```
    data.withUnsafeBytes { (pointer: UnsafePointer<UInt8>) in
      written = stream.write(pointer, maxLength: totalSize)
    }
```

Would have to be converted to this:

```
    data.withUnsafeBytes { (body: UnsafeRawBufferPointer) in
      if let baseAddress = body.baseAddress, body.count > 0 {
        let pointer = baseAddress.assumingMemoryBound(to: UInt8.self)
        written = stream.write(pointer, maxLength: totalSize)
      }
    }
```

But if someone tries to write this again, they might use the old way and forget to do the `count` check and the `baseAddress` check, leading to a crash. With the wrapper, we can make sure that no future additions will crash, and when we want to delete the old code we just remove the `#else` case in the wrapper.

In addition, if we want to convert all the call sites to the new API, we're going to have a problem because the new API in 5.0 is this:

```
func withUnsafeBytes<ResultType>(_ body: (UnsafeRawBufferPointer) throws -> ResultType) rethrows -> ResultType
```

The new function no longer is generic over the content type in the Data like the old API:

```
withUnsafeBytes<ResultType, ContentType>(_ body: (UnsafePointer<ContentType>) throws -> ResultType) rethrows -> ResultType
```

But in the implementation of the closure, you need to know the content type so you can call `assumingMemoryBound(to:)`:

```
let pointer = baseAddress.assumingMemoryBound(to: UInt8.self)
```

I can't find a way to provide that type to `assumingMemoryBound(to:)` at runtime, it needs to be provided at compile time. Either we use the old API approach of genericizing over `ResultType` and `ContentType` (which defeats the purpose of making this change), or we take in the type in the function:

```
internal func withUnsafeBytes<T, ResultType>(innerType: T.Type, _ body: (UnsafeRawBufferPointer) throws -> ResultType) rethrows -> ResultType {
    if count > 0 {
      return try withUnsafeBytes { (pointer: UnsafePointer<T>) -> ResultType in
        return try body(UnsafeRawBufferPointer(start: pointer, count: count))
      }
    }
  }
```

But then we're still not providing the exact same new API to older versions of Swift. So perhaps my approach is the only one that works for now, even though it will require changing the call sites when we remove `>=4.1` support.

But I'll defer to your preference on this, happy to do either approach.

tbkka commented on Sources/SwiftProtobuf/JSONEncoder.swift:
```diff
@@ -70,7 +70,11 @@ internal struct JSONEncoder {
 
     internal init() {}
 
+    #if swift(>=5.0)
```
As above, if the new syntax works in Swift 4, then we can avoid this conditional.

tbkka replied:
I just tried it; the new syntax does seem to work in Swift 4.

tbkka commented:
@alanzeino -- I'm testing a branch where I've globally replaced all of the `Data(bytes:)` with `Data(_:)`.  Once I finish running tests (for Swift 4.0 through 5.0), I'll put up a PR for that.  So I think you can focus on just the 24 warnings about `withUnsafeBytes` and `withUnsafeMutableBytes`.  If you can fix those, I think I've got the rest of the 5.0 issues covered.

alanzeino commented:
> @alanzeino -- I'm testing a branch where I've globally replaced all of the `Data(bytes:)` with `Data(_:)`. Once I finish running tests (for Swift 4.0 through 5.0), I'll put up a PR for that. So I think you can focus on just the 24 warnings about `withUnsafeBytes` and `withUnsafeMutableBytes`. If you can fix those, I think I've got the rest of the 5.0 issues covered.

Yep, I'll rebase.

keith commented:
@thomasvl @tbkka think we could merge this so there can be a new release that supports Swift 5?

tbkka commented:
The following seems sufficient to emulate the Swift 5 API on Swift 4:
```
#if !swift(>=5.0)
extension Data {
  func withUnsafeBytes<T>(_ body: (UnsafeRawBufferPointer) throws -> T) rethrows -> T {
    let c = count
    return try withUnsafeBytes { (p: UnsafePointer<UInt8>) throws -> T in
      try body(UnsafeRawBufferPointer(start: p, count: c))
    }
  }
  mutating func withUnsafeMutableBytes<T>(_ body: (UnsafeMutableRawBufferPointer) throws -> T) rethrows -> T {
    let c = count
    return try withUnsafeMutableBytes { (p: UnsafeMutablePointer<UInt8>) throws -> T in
      try body(UnsafeMutableRawBufferPointer(start: p, count: c))
    }
  }
}
#endif
```

Note that every place we used the old `withUnsafeBytes` APIs, we always used it with `UInt8`, so this is actually correct for us.

Using the above, I've experimentally switched a couple of places to use the Swift 5 APIs -- the result seems to work correctly on both Swift 4 and Swift 5.  In the process, I did see some spots where the resulting code could probably be simplified, but I think such improvements can wait until after we get a clean build.

alanzeino commented:
> The following seems sufficient to emulate the Swift 5 API on Swift 4:
> 
> ```
> #if !swift(>=5.0)
> extension Data {
>   func withUnsafeBytes<T>(_ body: (UnsafeRawBufferPointer) throws -> T) rethrows -> T {
>     let c = count
>     return try withUnsafeBytes { (p: UnsafePointer<UInt8>) throws -> T in
>       try body(UnsafeRawBufferPointer(start: p, count: c))
>     }
>   }
>   mutating func withUnsafeMutableBytes<T>(_ body: (UnsafeMutableRawBufferPointer) throws -> T) rethrows -> T {
>     let c = count
>     return try withUnsafeMutableBytes { (p: UnsafeMutablePointer<UInt8>) throws -> T in
>       try body(UnsafeMutableRawBufferPointer(start: p, count: c))
>     }
>   }
> }
> #endif
> ```
> Note that every place we used the old `withUnsafeBytes` APIs, we always used it with `UInt8`, so this is actually correct for us.
> 
> Using the above, I've experimentally switched a couple of places to use the Swift 5 APIs -- the result seems to work correctly on both Swift 4 and Swift 5. In the process, I did see some spots where the resulting code could probably be simplified, but I think such improvements can wait until after we get a clean build.

Ah I didn't think to hardcode the type. Updating now.

alanzeino commented:
Updated all the cases. One thing to mention, looks like the master branch does not compile. So I did this with a commit that did compile, then once I verified that it all compiled after my changes, I rebased so that the diff didn't have any conflicts.

Any plan to add a CI check to these PRs?

tbkka commented:
We don't have PR testing, but we do have CI for the master branch:

https://ci.swift.org/job/swift-protobuf-linux-ubuntu-16_04/

According to that, the master branch did build last night.  I'll take a look when I get a chance.

alanzeino commented:
> We don't have PR testing, but we do have CI for the master branch:
> 
> https://ci.swift.org/job/swift-protobuf-linux-ubuntu-16_04/
> 
> According to that, the master branch did build last night. I'll take a look when I get a chance.

If I were to guess, I'd say that master branch build uses `swift build` and not `xcodebuild build` — and the reason why this doesn't compile is because a new type (`TextFormatEncodingOptions`) was added in a new file, but the `.xcodeproj` wasn't updated to use that new file.

That also raises another question I had — should we add the .xcodeproj to the `.gitignore` and just generate it using Swift PM? Then we can add a CI job that does both `swift build` and then `swift package generate-xcodeproj && xcodebuild build -scheme SwiftProtobuf_iOS `, which would catch these project inconsistencies.

tbkka commented on Sources/SwiftProtobuf/AnyMessageStorage.swift:
```diff
@@ -66,22 +66,26 @@ fileprivate func unpack(contentJSON: Data,
+    if let baseAddress = body.baseAddress, body.count > 0 {
+      let bytes = baseAddress.assumingMemoryBound(to: UInt8.self)
+
+      let buffer = UnsafeBufferPointer(start: bytes, count: contentJSON.count)
```
Please use `body.count` here (the argument given to the closure) instead of `contentJSON.count` (which requires an additional capture)

tbkka commented on Sources/SwiftProtobuf/BinaryDecoder.swift:
```diff
@@ -1204,7 +1207,9 @@ internal struct BinaryDecoder: Decoder {
+                      if let baseAddress = body.baseAddress, body.count > 0 {
+                        let pointer = baseAddress.assumingMemoryBound(to: UInt8.self)
                         var extDecoder = BinaryDecoder(forReadingFrom: pointer,
                                                        count: data.count,
```
`data.count` ==> `body.count`

tbkka commented on Sources/SwiftProtobuf/Google_Protobuf_Any+Extensions.swift:
```diff
@@ -66,15 +66,19 @@ extension Google_Protobuf_Any {
+            var textDecoder = try TextFormatDecoder(
+              messageType: Google_Protobuf_Any.self,
+              utf8Pointer: bytes,
+              count: data.count,
```
`data.count` ==> `body.count`

tbkka commented on Sources/SwiftProtobuf/JSONEncoder.swift:
```diff
@@ -333,7 +333,10 @@ internal struct JSONEncoder {
+
                 var t: Int = 0
                 var bytesInGroup: Int = 0
                 for i in 0..<value.count {
```
`value.count` ==> `body.count`

tbkka commented on Sources/SwiftProtobuf/JSONScanner.swift:
```diff
@@ -848,15 +850,18 @@ internal struct JSONScanner {
+          (body: UnsafeRawBufferPointer) -> UInt64? in
+          if let baseAddress = body.baseAddress, body.count > 0 {
+            let bytes = baseAddress.assumingMemoryBound(to: UInt8.self)
+            let buffer = UnsafeBufferPointer(start: bytes, count: raw.count)
```
`raw.count` => `body.count`

tbkka commented on Sources/SwiftProtobuf/JSONScanner.swift:
```diff
@@ -907,15 +912,18 @@ internal struct JSONScanner {
+          (body: UnsafeRawBufferPointer) -> Int64? in
+          if let baseAddress = body.baseAddress, body.count > 0 {
+            let bytes = baseAddress.assumingMemoryBound(to: UInt8.self)
+            let buffer = UnsafeBufferPointer(start: bytes, count: raw.count)
```
`raw.count` => `body.count`

tbkka commented on Sources/SwiftProtobuf/JSONScanner.swift:
```diff
@@ -971,16 +979,19 @@ internal struct JSONScanner {
+            (body: UnsafeRawBufferPointer) -> Float? in
+            if let baseAddress = body.baseAddress, body.count > 0 {
+              let bytes = baseAddress.assumingMemoryBound(to: UInt8.self)
+              let buffer = UnsafeBufferPointer(start: bytes, count: raw.count)
```
`raw.count` ==> `body.count`

tbkka commented on Sources/SwiftProtobuf/JSONScanner.swift:
```diff
@@ -1042,15 +1053,18 @@ internal struct JSONScanner {
+            (body: UnsafeRawBufferPointer) -> Double? in
+            if let baseAddress = body.baseAddress, body.count > 0 {
+              let bytes = baseAddress.assumingMemoryBound(to: UInt8.self)
+              let buffer = UnsafeBufferPointer(start: bytes, count: raw.count)
```
`raw.count` ==> `body.count`

tbkka commented on Sources/SwiftProtobuf/Message+BinaryAdditions.swift:
```diff
@@ -104,12 +108,15 @@ extension Message {
+        if let baseAddress = body.baseAddress, body.count > 0 {
+          let pointer = baseAddress.assumingMemoryBound(to: UInt8.self)
+          var decoder = BinaryDecoder(forReadingFrom: pointer,
+                                      count: data.count,
```
`data.count` ==> `body.count`

tbkka commented on Sources/SwiftProtobuf/Message+JSONAdditions.swift:
```diff
@@ -89,21 +89,25 @@ extension Message {
+      if let baseAddress = body.baseAddress, body.count > 0 {
+        let bytes = baseAddress.assumingMemoryBound(to: UInt8.self)
+
+        let buffer = UnsafeBufferPointer(start: bytes, count: jsonUTF8Data.count)
```
`jsonUTF8Data.count` ==> `body.count`

tbkka commented on Sources/SwiftProtobuf/Message+JSONArrayAdditions.swift:
```diff
@@ -91,14 +91,19 @@ extension Message {
+
+      if let baseAddress = body.baseAddress, body.count > 0 {
+        let bytes = baseAddress.assumingMemoryBound(to: UInt8.self)
+        let buffer = UnsafeBufferPointer(start: bytes, count: jsonUTF8Data.count)
```
`jsonUTF8Data.count` ==> `body.count`

tbkka commented on Sources/SwiftProtobuf/Message+TextFormatAdditions.swift:
```diff
@@ -70,14 +70,17 @@ extension Message {
+            let bytes = baseAddress.assumingMemoryBound(to: UInt8.self)
+            var decoder = try TextFormatDecoder(messageType: Self.self,
+                                                utf8Pointer: bytes,
+                                                count: data.count,
```
`data.count` ==> `body.count`

tbkka commented on Sources/SwiftProtobuf/TextFormatEncoder.swift:
```diff
@@ -263,36 +263,40 @@ internal struct TextFormatEncoder {
+          if let baseAddress = body.baseAddress, body.count > 0 {
+            let p = baseAddress.assumingMemoryBound(to: UInt8.self)
+
             for i in 0..<value.count {
```
`value.count` ==> `body.count`

tbkka commented on Sources/SwiftProtobuf/TextFormatEncodingVisitor.swift:
```diff
@@ -77,11 +77,14 @@ internal struct TextFormatEncodingVisitor: Visitor {
+            if let baseAddress = body.baseAddress, body.count > 0 {
+              let p = baseAddress.assumingMemoryBound(to: UInt8.self)
               var decoder = BinaryDecoder(forReadingFrom: p,
                                           count: bytes.count,
```
`bytes.count` ==> `body.count`

tbkka commented on Sources/SwiftProtobuf/TextFormatEncodingVisitor.swift:
```diff
@@ -107,7 +110,10 @@ internal struct TextFormatEncodingVisitor: Visitor {
+                  let p = baseAddress.assumingMemoryBound(to: UInt8.self)
+
                   var testDecoder = BinaryDecoder(forReadingFrom: p,
                                                   count: bytes.count,
```
`bytes.count` ==> `body.count`

tbkka commented:
Does your branch pass the test suite?  (E.g. `make test`?)

I'm seeing failures here in the following three tests:
  `SwiftProtobufTests.Test_JSON testMultipleFields`
  `SwiftProtobufTests.Test_JSON_Array testTwoObjectsWithMultipleFields`
  `SwiftProtobufTests.Test_Map_JSON testMapInt32Bytes`

I find it interesting that the failing cases all relate to JSON and all of them have a `bytes` field which is empty.

thomasvl commented on SwiftProtobuf.xcodeproj/project.pbxproj:
```diff
@@ -288,6 +288,10 @@
 		DB2E0AFF1EB25D1D00F59319 /* Message+JSONArrayAdditions.swift in Sources */ = {isa = PBXBuildFile; fileRef = DB2E0AFD1EB25D1D00F59319 /* Message+JSONArrayAdditions.swift */; };
 		DB2E0B001EB25D1D00F59319 /* Message+JSONArrayAdditions.swift in Sources */ = {isa = PBXBuildFile; fileRef = DB2E0AFD1EB25D1D00F59319 /* Message+JSONArrayAdditions.swift */; };
 		DB2E0B011EB25D1D00F59319 /* Message+JSONArrayAdditions.swift in Sources */ = {isa = PBXBuildFile; fileRef = DB2E0AFD1EB25D1D00F59319 /* Message+JSONArrayAdditions.swift */; };
+		E7038A24224D7F1000B68775 /* Data+Extensions.swift in Sources */ = {isa = PBXBuildFile; fileRef = E7038A23224D7F1000B68775 /* Data+Extensions.swift */; };
```
Fixed this in #859, so when you update just make sure it isn't being added twice.

thomasvl replied:
Oops, fixed TextFormatEncodingOptions.swift in #859, you still need this change.

thomasvl commented:
> That also raises another question I had — should we add the `.xcodeproj` to the `.gitignore` and just generate it using Swift PM? Then we can add a CI job that does both swift build and then `swift package generate-xcodeproj && xcodebuild build -scheme SwiftProtobuf_iOS` , which would catch these project inconsistencies.

Since SwiftPM doesn't really do iOS, and to better support users wanting Carthage, etc.; I think we end up needing a project file checked in. 🙁 The `Makefile` does have a bunch of Xcode related targets (`test-xcode*`), but I don't think most people bother to check them often.

alanzeino commented:
> Does your branch pass the test suite? (E.g. `make test`?)
> 
> I'm seeing failures here in the following three tests:
> `SwiftProtobufTests.Test_JSON testMultipleFields`
> `SwiftProtobufTests.Test_JSON_Array testTwoObjectsWithMultipleFields`
> `SwiftProtobufTests.Test_Map_JSON testMapInt32Bytes`
> 
> I find it interesting that the failing cases all relate to JSON and all of them have a `bytes` field which is empty.

I should've taken your advice and just converted a few cases just to be safe. Unfortunately I have a bit of a problem in that I'm starting at a new employer next week that might not want me working on Open Source so how about for now I just land the back–ported extension in `Data+Extensions.swift` and convert one other place for reference, and I'll leave the rest to you since it would be a more careful strategy to do them piecemeal?

tbkka commented:
If you can fix the style issues I pointed out earlier, I'll see if I can debug that test failure.

alanzeino commented:
> If you can fix the style issues I pointed out earlier, I'll see if I can debug that test failure.

Okay!

tbkka commented on Sources/SwiftProtobuf/JSONScanner.swift:
```diff
@@ -190,8 +190,9 @@ private func parseBytes(
-        (dataPointer: UnsafeMutablePointer<UInt8>) in
-        var p = dataPointer
+        (body: UnsafeMutableRawBufferPointer) in
+      if let baseAddress = body.baseAddress, body.count > 0 {
```
The `value` here is actually the output buffer; the code here assumes that we're going to scan the input even if the output is zero-sized.  In particular, the close quote gets consumed deep inside the scan loop.  Since the code now skips that loop for zero-byte output, we end up not consuming the close quote for `""`.

I agree this is a somewhat confusing design.  I think the best fix would change how we handle the close quote:
* Look for two `case asciiDoubleQuote` below and remove `source.formIndex(after: &index)` at those points.
* Add `source.formIndex(after: &index)` just before `return value` at the end of the function.
This moves the close-quote handling outside of the `withUnsafeMutableBytes` which ensures that we skip it even if the output is empty.

alanzeino replied:
Ah, makes total sense. Making those changes now. Thanks!

tbkka commented:
Looks good.  I'm going to run some more tests before I merge this.

alanzeino commented:
Cool! Lemme know if there's more changes I should make.

tbkka commented:
I ran tests on Ubuntu with Swift 4.0 through a 5.0 beta and it all looks good.  Thanks for working through this!
